File: Javascript/src/models/CPMU.js
(function(exports) {

  function CPMU(){}

  CPMU.prototype.calculate = function(val1=null, val2=null){
    this.checkerror(val1, val2);
    return (parseFloat(val1) / parseFloat(val2) * 1000000).toFixed(8);
  };

  CPMU.prototype.quarter = function(val1=null){
    this.checkerror(val1);
    return (val1 / 3).toFixed(8);
  };

  CPMU.prototype.quartile = function(val1=null){
    this.checkerror(val1);
    var total = 4 - (val1 % 4);
    return total === 4 ? 1 : total + 1;
  };

  CPMU.prototype.display_date = function(val1=null){
    this.checkerror(val1);
    date = new Date(val1).toString().split(" ");
    return date[2] + " " + date[1] + " " + date[3];
  };

  CPMU.prototype.next_month = function(val1=null){
    this.checkerror(val1);
    return new Date(val1.setMonth(val1.getMonth() + 1));
  };

  CPMU.prototype.checkerror = function(val1=-1, val2=-1){
    if(val1===null || val2===null){
      throw new Error("Values not defined");
    }
  };

  exports.CPMU = CPMU;

})(this);

File: Ruby/app/models/CPMU.rb
class CPMU

  def self.calculate(val1=nil, val2=nil)
    check_error(val1, val2)
    val1.to_f / val2.to_f * 1000000
  end

  def self.quarter(val1=nil, val2=nil, val3=nil)
    check_error(val1, val2, val3)
    val1.to_f + val2.to_f + val3.to_f / 3
  end

  def self.quartile(val1=nil)
    check_error(val1)
    total = 4 - (val1 % 4)
    total == 4 ? 1 : total + 1
  end

  def self.display_date(val1=nil)
    check_error(val1)
    DateTime.parse(val1).strftime("%d %B %Y")
  end

  def self.next_month(val1=nil)
    check_error(val1)
    DateTime.parse(val1) >> 1
  end

  private

  def self.check_error(val1=-1, val2=-1, val3=-1)
    raise "Values not defined" if val1 == nil || val2 == nil || val3 == nil
  end

end
